### PR TITLE
Adds Runtime.NewNamespace to allow intentional name collisions

### PR DESCRIPTION
**Pull request by codefromthecrypt**
`Runtime.NewNamespace` creates an empty namespace which won't conflict with any other namespace including the default. This is more efficient than multiple runtimes, as namespaces share a compiler cache.

In simplest case, a namespace won't conflict if another has a module with the same name:
```go
	b := assemblyscript.NewBuilder(r)
	m1, _ := b.InstantiateModule(ctx, r.NewNamespace(ctx))
	m2, _ := b.InstantiateModule(ctx, r.NewNamespace(ctx))
```

This is also useful for different modules that import the same module name (like "env"), but need different
configuration or state. For example, one with trace logging enabled and another disabled:
```go
	b := assemblyscript.NewBuilder(r)

	// m1 has trace logging disabled
	ns1 := r.NewNamespace(ctx)
	_ = b.InstantiateModule(ctx, ns1)
	m1, _ := ns1.InstantiateModule(ctx, compiled, config)

	// m2 has trace logging enabled
	ns2 := r.NewNamespace(ctx)
	_ = b.WithTraceToStdout().InstantiateModule(ctx, ns2)
	m2, _ := ns2.InstantiateModule(ctx, compiled, config)
```

Note: to facilitate this there's one breaking change below:

`ModuleBuilder.Instantiate(context.Context)` -> `Instantiate(context.Context, Namespace)`

This also fixes a resource leak when modules implicitly compiled code (ex `Runtime.InstantiateModuleFromCode`).

See #599

### CODE REVIEW

File: builder.go [hunk 1]
```diff
@@ -302,11 +302,6 @@ func (b *moduleBuilder) Instantiate(ctx context.Context) (api.Module, error) {
 	if compiled, err := b.Compile(ctx, NewCompileConfig()); err != nil {
 		return nil, err
 	} else {
-		if err = b.r.store.Engine.CompileModule(ctx, compiled.(*compiledCode).module); err != nil {
```
Comment (codefromthecrypt) [inline builder.go:305]: this was redundant

**Comment by codefromthecrypt** [comment]
last step is adding an example. the example will show instantiate multiple stateful host modules under the same name using namespaces.

ex. each namespace includes "env" created from ModuleBuilder and a wasm source module that imports it. A function in "env" mutates its own map without function parameters needing to include a reference to it (externref or a uintptr). 

This should prove the isolation needed for #599 and also demonstrate how new sorts of module-scoped state not already in SysContext can be achieved without the use of go context (context.Context). cc @prep

**Comment by codefromthecrypt** [comment]
Last thing I'll do is backfill tests to show that when there's implicit compilation of a module, the compiled code is deleted when the module closes. This was unrelated but easy to fix while here.

File: internal/wasm/engine.go [hunk 2]
```diff
@@ -11,6 +11,14 @@ type Engine interface {
 	// CompileModule implements the same method as documented on wasm.Engine.
 	CompileModule(ctx context.Context, module *Module) error
 
+	// CompiledModuleCount is exported for testing, to track the size of the compilation cache.
```
Comment (mathetake) [inline internal/wasm/engine.go:14]: I see this is only used in tests, so maybe adds a comment about that?

**Review by mathetake — APPROVED** [review]

**Comment by codefromthecrypt** [comment]
thanks for taking a break from simd to have a look @mathetake .. now back to simd 🧑‍🚒 ;)

**Pull request merged**